keps/sig-api-machinery/5116-streaming-response-encoding/README.md

```diff
@@ -0,0 +1,354 @@
+Further details can be found in [kubernetes/kubernetes#129304](https://github.com/kubernetes/kubernetes/issues/129304#issuecomment-2612704644).
+
+### Risks and Mitigations
+
+
+## Design Details
+
+Implementing streaming encoders specifically for LIST responses significantly reduces the scope,
+allowing us to focus on a limited set of types and avoid the complexities of a fully generic streaming encoder.
+The core difference in our approach will be special handling of the `Items` field within LIST structures.
+Instead of encoding the entire `Items` array at once, we will iterate through the array and encode each item individually, streaming the encoded data to the client.
+
+This targeted approach enables the following implementation criteria:
+
+* **Strict Validation:**  Before proceeding with streaming encoding, 
+  the implementation will rigorously validate the Go struct tags of the target type. 
+  If the tags do not precisely match the expected structure, we will fallback to standard encoder.
+  This precautionary measure prevents incompatibility upon change of structure fields or encoded representation.
+* **Delegation to Standard Encoder:**  The encoding of all fields *other than* `Items`,
+  as well as the encoding of each individual item *within* the `Items` array,
+  will be delegated to the standard `encoding/json` (for JSON) or `protobuf` (for Proto) packages.
+  This leverages the existing, well-tested encoding logic and minimizes the amount of custom code required, reducing the risk of introducing bugs.
+
+The types requiring custom streaming encoders are:
+
+* `*List` types for built-in Kubernetes API resources (e.g., `PodList`, `ConfigMapList`).
+* `UnstructuredList` for Custom Resources.
+* `runtime.Unknown` used for Proto encoder to provide type information.
+
+To further enhance robustness, a static analysis check will be introduced to detect and prevent any inconsistencies in Go struct tags across different `*List` types.
+This addresses the concern that not all `*List` types may have perfectly consistent tag definitions.
+
+### Gzip encoding
+
+As pointed out in kubernetes/kubernetes#129334#discussion_r1938405782,
+the current Kubernetes gzip encoding implementation assumes the response is written in a single large chunk,
+checking just first write size to determine if the response is large enough for compression.
+This is a bad assumption about internal encoder implementation details and should be fixed regardless.
+
+To ensure gzip compression works well with streaming,
+we will preempt all encoder changes by fixing the gzip compression.
+First will add unit tests that will prevent subsequent changes from impacting results,
+especially around the compression threshold.
+Then, we will rewrite the gzip compression to buffer the response and delay the
+decision to enable compression until we have observed enough bytes to hit the threshold
+or we received whole response and we can write it without compressing.
```

Comment:
just an implementation note, by adding an intermediate buffer to the gzip writer wrapper, be careful not to reintroduce the issue where we don't flush headers until a non-zero number of bytes are written

https://github.com/kubernetes/kubernetes/pull/31446/files#diff-67dd9e8f3ee257072765326cb4f242852554a2c0753563fa51e292c0a63a7b94R2488

maybe the code path that hits deferredResponseWriter isn't vulnerable to that issue if it already has the single object it is serializing, but watch out for buffering a generic gzip writer wrapper if we're not sure whether or not we'll have immediate output